File: services/nuldoc/nuldoc-src/pages/SlidePage.ts
import GlobalFooter from "../components/GlobalFooter.ts";
import GlobalHeader from "../components/SlidesGlobalHeader.ts";
import PageLayout from "../components/PageLayout.ts";
import StaticScript from "../components/StaticScript.ts";
import { Config, getTagLabel } from "../config.ts";
import { dateToString } from "../revision.ts";
import { Slide } from "../slide/slide.ts";
import { getPostPublishedDate } from "../generators/post.ts";
import {
  a,
  article,
  button,
  div,
  elem,
  Element,
  h1,
  h2,
  header,
  li,
  ol,
  section,
  ul,
} from "../dom.ts";

export default async function SlidePage(
  slide: Slide,
  config: Config,
): Promise<Element> {
  return await PageLayout({
    metaCopyrightYear: getPostPublishedDate(slide).year,
    metaDescription: slide.title,
    metaKeywords: slide.tags.map((slug) => getTagLabel(config, slug)),
    metaTitle:
      `${slide.event} (${slide.talkType})｜${config.sites.slides.siteName}`,
    requiresSyntaxHighlight: true,
    site: "slides",
    config,
    children: elem(
      "body",
      { class: "single" },
      GlobalHeader({ config }),
      elem(
        "main",
        { class: "main" },
        article(
          { class: "post-single" },
          header(
            { class: "post-header" },
            h1({ class: "post-title" }, slide.title),
            slide.tags.length !== 0
              ? ul(
                { class: "post-tags" },
                ...slide.tags.map((slug) =>
                  li(
                    { class: "tag" },
                    a(
                      { href: `/tags/${slug}/` },
                      getTagLabel(config, slug),
                    ),
                  )
                ),
              )
              : null,
          ),
          div(
            { class: "post-content" },
            section(
              { id: "changelog" },
              h2({}, a({ href: "#changelog" }, "更新履歴")),
              ol(
                {},
                ...slide.revisions.map((rev) =>
                  li(
                    { class: "revision" },
                    elem(
                      "time",
                      { datetime: dateToString(rev.date) },
                      dateToString(rev.date),
                    ),
                    `: ${rev.remark}`,
                  )
                ),
              ),
            ),
            elem("canvas", { id: "slide", "data-slide-link": slide.slideLink }),
            div(
              {},
              button({ id: "prev", type: "button" }, "Prev"),
              button({ id: "next", type: "button" }, "Next"),
            ),
            await StaticScript({
              site: "slides",
              fileName: "/slide.js",
              type: "module",
              config,
            }),
          ),
        ),
      ),
      GlobalFooter({ config }),
    ),
  });
}

File: services/nuldoc/lib/nuldoc/pages/slide_page.rb
module Nuldoc
  module Pages
    class SlidePage
      def initialize(slide:, config:)
        @slide = slide
        @config = config
      end

      def render
        slide = @slide
        config = @config
        Components::PageLayout.new(
          meta_copyright_year: GeneratorUtils.published_date(slide).year,
          meta_description: "「#{slide.title}」(#{slide.event} で登壇)",
          meta_keywords: slide.tags.map { |slug| config.tag_label(slug) },
          meta_title: "#{slide.title} (#{slide.event})｜#{config.sites.slides.site_name}",
          site: 'slides',
          config: config,
          children: DOM::HTMLBuilder.new.build do
            body class: 'single' do
              render Components::StaticStylesheet, site: 'slides', file_name: '/slides.css', config: config
              render Components::SlidesGlobalHeader, config: config
              main class: 'main' do
                article class: 'post-single' do
                  header class: 'post-header' do
                    h1(class: 'post-title') { text slide.title }
                    if slide.tags.length.positive?
                      ul class: 'post-tags' do
                        slide.tags.each do |slug|
                          li class: 'tag' do
                            a(class: 'tag-inner', href: "/tags/#{slug}/") { text config.tag_label(slug) }
                          end
                        end
                      end
                    end
                  end
                  div class: 'post-content' do
                    section id: 'changelog' do
                      h2 { a(href: '#changelog') { text '更新履歴' } }
                      ol do
                        slide.revisions.each do |rev|
                          ds = Revision.date_to_string(rev.date)
                          li class: 'revision' do
                            time(datetime: ds) { text ds }
                            text ": #{rev.remark}"
                          end
                        end
                      end
                    end
                    div class: 'slide-container', id: 'slide-container',
                        'data-slide-link': slide.slide_link do
                      div class: 'slide-stage', id: 'slide-stage' do
                        canvas id: 'slide'
                        div class: 'textLayer', id: 'text-layer'
                        div class: 'annotationLayer', id: 'annotation-layer'
                      end
                      div class: 'slide-status', id: 'slide-status',
                          role: 'status', 'aria-live': 'polite', hidden: 'hidden'
                    end
                    noscript do
                      style { raw '.slide-container,.controllers-buttons{display:none}' }
                      p class: 'slide-noscript' do
                        a(href: slide.slide_link) { text 'PDF をダウンロード' }
                      end
                    end
                    div class: 'controllers' do
                      div class: 'controllers-buttons' do
                        button id: 'prev', type: 'button',
                               'aria-label': '前のページ (←/h)', title: '前のページ (←/h)' do
                          elem 'svg', width: '20', height: '20', viewBox: '0 0 24 24', fill: 'none',
                                      stroke: 'currentColor', 'stroke-width': '2' do
                            elem 'path', d: 'M15 18l-6-6 6-6'
                          end
                        end
                        div class: 'page-indicator' do
                          input id: 'page-input', class: 'page-input', type: 'number',
                                min: '1', step: '1', value: '1', inputmode: 'numeric',
                                'aria-label': 'ページ番号'
                          span(class: 'page-separator') { text '/' }
                          span(class: 'page-count', id: 'page-count') { text '–' }
                        end
                        button id: 'next', type: 'button',
                               'aria-label': '次のページ (→/l)', title: '次のページ (→/l)' do
                          elem 'svg', width: '20', height: '20', viewBox: '0 0 24 24', fill: 'none',
                                      stroke: 'currentColor', 'stroke-width': '2' do
                            elem 'path', d: 'M9 18l6-6-6-6'
                          end
                        end
                      end
                      a(class: 'slide-download', href: slide.slide_link, download: '') do
                        text 'PDF をダウンロード'
                      end
                    end
                    render Components::StaticScript,
                           site: 'slides',
                           file_name: '/slide.js',
                           type: 'module',
                           config: config
                  end
                end
              end
              render Components::GlobalFooter, config: config
            end
          end
        ).build
      end
    end
  end
end
